### PR TITLE
Continue String to Symbol conversion in rustdoc (1)

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Follow-up of https://github.com/rust-lang/rust/pull/80091.

This PR is already big enough so I'll stop here before the next one.

r? @jyn514

### CODE REVIEW

File: src/librustdoc/html/format.rs [hunk 1]
```diff
@@ -1160,7 +1159,7 @@ impl clean::Import {
     crate fn print(&self) -> impl fmt::Display + '_ {
         display_fn(move |f| match self.kind {
             clean::ImportKind::Simple(ref name) => {
-                if *name == self.source.path.last_name() {
+                if name.as_str() == self.source.path.last_name() {
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/format.rs:1162]: Realizing that now it should be instead: `if name == self.source.path.last() {`. Updating once confirmed.

File: src/librustdoc/html/format.rs [hunk 2]
```diff
@@ -445,11 +445,10 @@ impl clean::GenericArgs {
 impl clean::PathSegment {
     crate fn print(&self) -> impl fmt::Display + '_ {
         display_fn(move |f| {
-            f.write_str(&self.name)?;
             if f.alternate() {
-                write!(f, "{:#}", self.args.print())
+                write!(f, "{}{:#}", self.name.as_str(), self.args.print())
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/format.rs:449]: No need for the `.as_str()`.

File: src/librustdoc/html/format.rs [hunk 3]
```diff
@@ -445,11 +445,10 @@ impl clean::GenericArgs {
...
             } else {
-                write!(f, "{}", self.args.print())
+                write!(f, "{}{}", self.name.as_str(), self.args.print())
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/format.rs:451]: No need for the `.as_str()`.

File: src/librustdoc/clean/types.rs [hunk 4]
```diff
@@ -1407,6 +1405,36 @@ impl PrimitiveType {
         }
     }
 
+    crate fn get_symbol(self) -> Symbol {
```
Comment (jyn514) [inline src/librustdoc/clean/types.rs:1408]: Heh, I also had to add this for https://github.com/rust-lang/rust/pull/80090. Can you call it `as_sym` for consistency with `as_str`?

File: src/librustdoc/clean/types.rs [hunk 5]
```diff
@@ -1675,13 +1703,17 @@ crate struct Path {
...
+    }
+
+    crate fn last_name(&self) -> SymbolStr {
         self.segments.last().expect("segments were empty").name.as_str()
     }
 
     crate fn whole_name(&self) -> String {
         String::from(if self.global { "::" } else { "" })
-            + &self.segments.iter().map(|s| s.name.clone()).collect::<Vec<_>>().join("::")
+            + &self.segments.iter().map(|s| s.name.to_string()).collect::<Vec<_>>().join("::")
```
Comment (jyn514) [inline src/librustdoc/clean/types.rs:1716]: Oof, this makes me sad :( but it's not any worse than the existing code so no need to fix it here.

File: src/librustdoc/json/conversions.rs [hunk 6]
```diff
@@ -163,7 +163,9 @@ impl From<clean::ItemKind> for ItemEnum {
         use clean::ItemKind::*;
         match item {
             ModuleItem(m) => ItemEnum::ModuleItem(m.into()),
```
Comment (jyn514) [inline src/librustdoc/json/conversions.rs:165]: ItemEnum should probably be renamed at some point, but no need to do it here.

**Review by jyn514 — COMMENTED** [review]
This also looks great :heart:

File: src/librustdoc/html/format.rs (hunk 1 above)
Comment (jyn514) [reply src/librustdoc/html/format.rs]: What's the difference between the two?

File: src/librustdoc/clean/types.rs (hunk 4 above)
Comment (GuillaumeGomez) [reply src/librustdoc/clean/types.rs:1408]: Sure!

File: src/librustdoc/json/conversions.rs (hunk 6 above)
Comment (GuillaumeGomez) [reply src/librustdoc/json/conversions.rs:165]: Maybe open an issue so we don't forget? (And mark it as `e-easy` too ;) )

**Comment by jyn514** [comment]
-2.2% on instruction count, -1.5% on max-rss :tada: :tada: 

@bors r+ rollup=never

File: src/librustdoc/html/format.rs (hunk 1 above)
Comment (GuillaumeGomez) [reply src/librustdoc/html/format.rs]: In the original code I was comparing strings whereas in the second we compare `Symbol` (so usize). I assume the second one to be faster. ;)

**Pull request merged**